Speedup static content deploy for specific theme

ihor-sviziev opened the pull request:
### Description (*)
This PR adds improvement for ability to deploy the only specific theme(s) for multiple locales, so it should improve build phase of deployment for the those who use single theme with different locales.

Basically what I did - i just removed packages from all themes from which inherited requested theme, so in case of Luma - we have only 1 parent theme, but for people who has bigger inheritance - different will be bigger.

How I tested:

**Before:**
```bash
php bin/magento setup:upgrade
php bin/magento setup:static-content:deploy -f -t Magento/luma -- en_AU en_CA en_IE en_NZ en_GB en_US
```
~114 seconds ❌
![image](https://user-images.githubusercontent.com/1873745/97312245-f07a9780-186d-11eb-85bb-97ada5723fee.png)



**After:**
```bash
php bin/magento setup:upgrade
php bin/magento setup:static-content:deploy -f -t Magento/luma --no-parent -- en_AU en_CA en_IE en_NZ en_GB en_US
```
~60 seconds ✔
![image](https://user-images.githubusercontent.com/1873745/97312501-33d50600-186e-11eb-9d7c-c6d82fa2e4aa.png)



### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. N/A

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#30184: Speedup static content deploy for specific theme

ihor-sviziev committed:
Speedup static content deploy for specific theme

ihor-sviziev commented on app/code/Magento/Deploy/Strategy/QuickDeploy.php:
```diff
@@ -131,4 +136,28 @@ private function getInheritanceLevel(Package $package)
+     */
+    private function canDeployTheme(string $theme, array $includedThemesMap, array $excludedEntitiesMap): bool
+    {
+        $includesAllThemes = \array_key_exists('all', $includedThemesMap);
```
Using here `\array_key_exists` instead of `isset` because of optimization https://github.com/php/php-src/pull/3360 that was done in php 7.4, and definitely not in_array because of performance issues.

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

hostep commented:
Nice! So this will revert the SCD behavior back to how it was in Magento 2.1 from memory.

It might cause performance decreases for deploying multiple themes which inherit from the same parent theme (since it might not generate the parent theme files and can't re-use them), but I'm not sure. If that's the case, you can probably first deploy the parent theme followed by the child themes, that should result in the same performance probably.

I'm assuming you tested that the resulting generated files are 100% the same as before this change?

ihor-sviziev commented:
@hostep i didn’t find any difference in result and in performance between generating multiple themes, or single theme

hostep commented:
Cool!

Regarding performance, I'm curious if there will be a difference if we have a theme structure like:
```
parent theme
 - child theme 1
 - child theme 2
 - child theme 3
```

Also: should the same change be added to the `StandardDeploy` and `CompactDeploy` classes?

ihor-sviziev commented:
@hostep actually I never used them, might check that if needed, but not sure if it will work for them.

mrtuvn commented:
Maybe not related in scope of this but i'm just curious about the use case magento site setup multiple site (around 15 theme included magento theme base) and various some locales (2-3 locales). When owner run deploy all theme for all locales! No exclude

ihor-sviziev commented:
@mrtuvn in this case it will work as it was before ✅

BarnyShergold commented:
How will this affect the build phase for Magento Cloud where we can control what themes are built and what locales? Or will it not have an effect?

ihor-sviziev commented:
@BarnyShergold unfortunately I don’t have access to ee and magento cloud so can’t check it, but I think it should work fine as it’s basically runs magento static content deploy command with specified list of themes and locales, the only change will be - the only requested themes will be compiled

ihor-sviziev commented:
@magento create issue

NadiyaS commented:
Hi,
There are some results after testing this implementation on Cloud integration environment.
`SCD_MATRIX` configuration was used to test SCD with separate CLI command for themes.
Generation of `luma` theme was faster from ~ 17% **up to 45%** (depends on configurations in `SCD_MATRIX`). 
E.g. 
```stage:
  build:
    SCD_MATRIX:
      Magento/blank:
        language:
          - fr_CA
          - fr_FR
          - nl_NL
          - de_DE
          - en_CA
          - en_US
      Magento/luma:
        language:
          - en_US
          - en_AU
          - en_IE 
          - en_NZ 
          - en_GB
          - en_CA
```

For this configuration `ece-tools` runs 3 CLI commands:
- `php ./bin/magento setup:static-content:deploy --ansi --no-interaction -f --jobs 4 --no-html-minify --exclude-theme Magento/luma --exclude-theme Magento/blank en_US en_AU en_IE en_NZ en_GB en_CA`
- `php ./bin/magento setup:static-content:deploy --ansi --no-interaction -f --jobs 4 --no-html-minify --theme Magento/blank fr_CA fr_FR nl_NL de_DE en_CA en_US`
- `php ./bin/magento setup:static-content:deploy --ansi --no-interaction -f --jobs 4 --no-html-minify --theme Magento/luma en_US en_AU en_IE en_NZ en_GB en_CA`
Time for first 2 commands took ~ 33 sec and 22 sec accordingly (as for new as for old implementation)
Generation of `luma` theme took **40 sec** before changes and **25 sec** after
Generation of all static content for all themes was on ~ 17% faster.

ihor-sviziev commented:
Open questions:

### Backward compatibility

Currently someone could depends on parent themes, so for these people this change.

**Possible solutions:**
- add new optional param, for instance `--no-parent` to `bin/magento setup:static-content:deploy` command
- merge this change to 2.5
- ...any other options? 

### Different behavior between the `quick`, `compact` and `standard` strategies

Currently my fix is improving only `quick` strategy, but BTH I never used any other strategies and don't really know if there any reason to use them.

From my perspective - it's better to support only one strategy and have this fix already merged.

YPyltiai commented:
@ihor-sviziev , as discussed on Slack - option should be good to have to get this change into earlier versions. We can then adjust ECE-Tools to handle this flag and make it configurable through current configs. If this is confirmed working well - in 2.5 what flag does could be adjusted to become a default behavior.

hostep commented:
I agree with the new `--no-parent` flag.

Regarding the different strategies, I would strongly recommend to try to apply the same change to the `standard` strategy, I think it's going to work (even though very few people use it probably). At the very least, we should make an attempt at trying this, if it fails and it causes too much headaches to get it sorted out, we can chose to not apply it.

The `compact` strategy is used more often (because it runs [many times faster](https://github.com/baldwin-agency/magento2-module-less-js-compiler#results) then `standard` or `quick`, but the downside is that it adds extra javascript overhead on the frontend of the shop and that certain 3rd party modules will break when using it).
But, I'm not sure if the same change can be applied due to how it works, it might need the parent theme to be deployed as well, but I'm not sure...

Also: if the change doesn't apply to one or two strategies, the description of this new `--no-parent` flag should highlight that it only affects certain strategies.

ihor-sviziev commented:
Ok, thx for the info, probably will update my pr by the end of week

ihor-sviziev commented:
@hostep does this pr fixes https://github.com/magento/magento2/issues/10674?

hostep commented:
> @hostep does this pr fixes #10674?

Hehe, I thought about that as well, if you don't change the Standard strategy (or if it's impossible to change it), then it fixes it yes 😛

The problem is that there is no public information available about how the standard and quick strategies should be different from each other, there is probably an internal architectural document with Magento somewhere. But us outsiders have no idea where it went wrong with these strategies when they implemented them in 2.2.0

orlangur commented:
@NadiyaS @YPyltiai could you please approve this PR? Feel free to unassign me then

YPyltiai commented:
From Product side the change looks good to me once suggestions in this thread are implemented. I will leave it up to @NadiyaS to review the code though.

ihor-sviziev commented:
I did following things:
1. Added new `--no-parent` optional parameter to the `setup:static-content:deploy` command
2. Added validation for `--no-parent` flag when used with any other strategy (will fail with message `Argument "no-parent" is not supported with "compact" strategy`)
3. Small cleanup - re-order options, use strict comparison, etc.

ihor-sviziev commented:
@magento run all tests

hostep commented:
@ihor-sviziev : while you're working on this, could you quickly check if the same optimisation could be applied to the StandardDeploy please? Then at least we know if it's possible or not?

hostep commented:
Sorry, my comment came too soon, I didn't saw your comment about:
> Also I checked - seems like it will be not so easy to add support of --no-parent argument to other strategies

Nevermind then!
You can now probably add https://github.com/magento/magento2/issues/10674 to the fixed issues list 😉 
You can add

ihor-sviziev commented:
@hostep basically I looked at compact strategy, it's not actually possible. Seems like I was able to do the same improvement for standard strategy. Adding my changes...

ihor-sviziev committed:
Speedup static content deploy for specific theme

Add support for standard strategy

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented on app/code/Magento/Deploy/Console/DeployStaticOptions.php:
```diff
@@ -127,6 +127,11 @@ class DeployStaticOptions
+    /**
+     * Key for not compiling parent themes
+     */
+    const NO_PARENT = 'no-parent';
```
This probably backward incompatible change, but this class insn't marked as `@api` and that's the most logical place to put this new argument name

ihor-sviziev commented:
@magento run Functional Tests B2B

ihor-sviziev commented:
Some notice should be fixed in integration test

ihor-sviziev committed:
Speedup static content deploy for specific theme

Fix failing integration test

ihor-sviziev commented:
@magento run Integration Tests

ihor-sviziev commented:
@magento run Static Tests

ihor-sviziev commented:
@magento run all tests

NadiyaS commented:
@ihor-sviziev This PR is approved by Magento Cloud and may continue to go through the standard approval process.

ihor-sviziev commented:
@magento run Integration Tests

ihor-sviziev commented:
Hi @orlangur could you review and approve my PR?

engcom-Bravo commented:
:heavy_check_mark: QA Passed

Perform 
`php bin/magento setup:upgrade`
`php bin/magento setup:static-content:deploy -f -t Magento/luma --no-parent -- en_AU en_CA en_IE en_NZ en_GB en_US`
**Before** applying changes provided in the PR, content deploy with `--no-parent` parameter was not available
![noparrent](https://user-images.githubusercontent.com/51680745/97977857-7ac67c80-1dd5-11eb-91c9-e007d7604afc.png)
Performing `php bin/magento setup:static-content:deploy -f -t Magento/luma -- en_AU en_CA en_IE en_NZ en_GB en_US` finished in 269 seconds
![Before](https://user-images.githubusercontent.com/51680745/97978010-ad707500-1dd5-11eb-89d8-781ab1a0da49.png)

**After** switching on the PR, the `php bin/magento setup:static-content:deploy -f -t Magento/luma --no-parent -- en_AU en_CA en_IE en_NZ en_GB en_US` became available and finished faster 
![After](https://user-images.githubusercontent.com/51680745/97978259-fb857880-1dd5-11eb-8f39-eed13f5e822c.png)

Commit:
Merge branch '2.4-develop' into improve-static-content-deploy-for-specific-theme

gabrieldagama commented:
@magento run all tests

ihor-sviziev commented:
@magento run Functional Tests B2B

ihor-sviziev commented:
@magento run Functional Tests B2B

@gabrieldagama any reasons why this PR isn't getting merged?

gabrieldagama commented:
@ihor-sviziev we were having some failed tests that I had suspected could be related to it. I'm trying to deliver it again, hopefully, we will merge it soon.

mrtuvn commented:
Hi just curiously how fast this changes brought with specific case we have multiple theme + multiple locale in compact deploy mode  in magento cloud? Anyone have benchmark this case
CC: @ihor-sviziev @gabrieldagama

hostep commented:
@mrtuvn: I can't speak about cloud, but just about Magento OS 2.4.2-p1 on premise.

Last week I did some optimisations to a certain project's deploy flow where we went from 183.5 seconds to 36.5 seconds (only for the SCD commands that is). This was using a 2 different frontend themes that both had two levels of parent themes, 4 different locale's got deployed for the frontend and the Magento backend theme only got deployed in one locale.

We accomplished this by:
- using 4 threads instead of 1 (`--jobs` flag)
- using the new `--no-parent` flag introduced here (again, very big thank you to @ihor-sviziev for this!)
- also used the `--no-js-bundle` flag, because we don't use js bundles on this particular project (this does shave of a few seconds believe it or not)
- removing the `web/tailwind` directory in a custom frontend theme, between compiling tailwind css and running the Magento SCD command (this particular project uses [Hyvä frontend technology](https://hyva.io/), which uses [tailwind](https://tailwindcss.com/))

Not sure if these are really helpful numbers, because of the particularities of this project, but it's something 🙂

mrtuvn commented:
thanks you very much for useful infor

mrtuvn commented:
Btw seem this fix not delivered to 2.3.7! Can it work for 2.3

ihor-sviziev commented:
@mrtuvn, we just applied changes from this PR as a patch and using it in 2.3 release line. 👍

mrtuvn commented:
@magento convert to patch for 2.3.7

m2-github-services commented:
@mrtuvn changes has been successfully converted for 2.3.7 version and pushed to the  converted-magento-magento2-30090-2.3.7 branch. 

 You may submit a pull request using the link: https://github.com/magento/magento2/compare/2.3.7-release...magento:converted-magento-magento2-30090-2.3.7

HenKun commented:
https://github.com/magento convert to patch for 2.4.1

HenKun commented:
@magento convert to patch for 2.4.1

m2-github-services commented:
@HenKun an error occurred during the patch generation.
Failure can be caused by the incompatible with base branch changes.
To replicate the issue try to apply patch locally using `git am` command.
